proof/crefine/AARCH64/Fastpath_C.thy

```diff
@@ -272,221 +258,223 @@ lemma lookup_fp_ccorres':
+          apply (rule_tac P="valid_cap' acap" in ccorres_from_vcg[where P'=UNIV])
+          apply (rule allI, rule conseqPre, vcg)
+          apply (clarsimp simp: valid_cap'_def isCap_simps if_1_0_0)
+          apply (simp add: unat_eq_0[symmetric] whenE_def returnOk_def return_def)
          apply (rule ccorres_cutMon)
-         apply (simp add: cutMon_walk_if cong: call_ignore_cong)
-         apply (rule_tac Q'="\<lambda>s. ret__int_' s = from_bool (isCNodeCap rv)"
-                       in ccorres_cond_both'[where Q=\<top>])
-           apply (clarsimp simp: from_bool_0)
-          apply (rule ccorres_rhs_assoc)+
-          apply (rule_tac P="ccorres r xf Gd Gd' hs a" for r xf Gd Gd' hs a in rsubst)
-           apply (rule "1.hyps",
-                  (rule refl in_returns in_bind[THEN iffD2, OF exI, OF exI, OF conjI]
-                       acc_CNodeCap_repr
-                        | assumption
-                        | clarsimp simp: unlessE_whenE locateSlot_conv
-                                         "1.prems"
-                        | clarsimp simp: whenE_def[where P=False])+)[1]
-          apply (simp add: whileAnno_def extra_sle_sless_unfolds)
-         apply (rule ccorres_drop_cutMon)
-         apply (simp add: liftE_def getSlotCap_def)
-         apply (rule ccorres_pre_getCTE)
-         apply (rule ccorres_cond_false_seq)
-         apply (rule_tac P="\<lambda>s. cteCap rva = rv" and P'="{s. cap_' s = cap}"
-                      in ccorres_from_vcg_throws)
-         apply (rule allI, rule conseqPre, vcg)
-         apply (clarsimp simp: return_def returnOk_def word_sle_def isRight_def)
-        apply simp
-        apply (wp getSlotCap_wp)
-       apply (simp add: if_1_0_0)
+         apply (simp add: liftE_bindE locateSlot_conv
+                     del: Collect_const)
+         apply (rule_tac P="abits = capCNodeBits acap + capCNodeGuardSize acap"
+                     in ccorres_case_bools2)
+          apply (rule ccorres_drop_cutMon)
+          apply (simp del: Collect_const)
+          apply (simp add: liftE_def getSlotCap_def del: Collect_const)
+          apply (rule ccorres_Guard_Seq)+
+          apply csymbr+
+          apply simp
+          apply (rule ccorres_move_array_assertion_cnode_ctes
+                      ccorres_move_c_guard_cte
+                      ccorres_rhs_assoc | csymbr)+
+          apply (rule getCTE_cap_h_val_ccorres_split)
+           apply ceqv
+          apply (rename_tac "getCTE_cap")
+          apply (csymbr | rule ccorres_Guard_Seq)+
+          apply (rule ccorres_cond_false_seq)
+          apply (simp add: ccorres_expand_while_iff_Seq[symmetric]
+                           whileAnno_def del: Collect_const)
+          apply (rule ccorres_cond_false)
+          apply (rule ccorres_cond_false_seq)
+          apply (simp del: Collect_const)
+          apply (rule_tac P'="{s. cap_' s = getCTE_cap}"
+                         in ccorres_from_vcg_throws[where P=\<top>])
```

Comment:
What's our style policy for the `in`? Should this be the following?
```suggestion
          apply (rule_tac P'="{s. cap_' s = getCTE_cap}"
                          in ccorres_from_vcg_throws[where P=\<top>])
```

Reply:
I asked about this in mattermost before I did these, and people linked me to the discussion where we had 3 acceptable versions I think (rhs `and`: the `in` is like you point out; lhs `and` with `in` indented same as the `and`; the `and/in` right-aligned with the `c` in `rule_tac` (Michael style?)). I'm a bit confused at this point. I was confused before as well, because I started the original discussion.
Worse yet, trying to align the `in` so that the `n` is under the `c`, when seen in a normal text editor (or diff), results in the `n` being under the space instead.
Advice welcome.

Reply:
Hmm, I guess our original discussion only considered the case with multiple instantiations. Sorry to add yet another version, but I thought our preference from that discussion was what I think you're describing with the right-aligned Michael style. That would make it
```suggestion
          apply (rule_tac P'="{s. cap_' s = getCTE_cap}"
                       in ccorres_from_vcg_throws[where P=⊤])
```

Reply:
Ok, I'm glad I'm not the only one who is confused :-)

I propose we leave this one as is and figure out separately what we really want for this case.

Reply:
This looks like I'm carrying some kind of rendering issue with even placing the `in` in the right place, because the *intended* placement was where @corlewis said it should be. I'll fix those misalignments with vim for now so it's like what Corey said, and then I'll create an issue so we sort this out and can put it in the style guide.

Reply:
OK it seems in the final ccorres proofs I wasn't making the misalignment mistakes, so maybe they didn't come from me. It's still not entirely consistent w.r.t. the `in` placement so TODO: make issue.
Fix pushed.